### PR TITLE
JWWEB-331 Implementation of Receive Digital Assets page

### PR DESCRIPTION
**Pull request by Creepypoke**
Receive digital assets page was implemented

### CODE REVIEW

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import React, { Component } from 'react'
+import { Link } from 'react-router'
+
+import { JText, JIcon, JRaisedButton } from 'components/base'
+import { AddressPicker } from 'components'
+import { clipboard, qrCode } from 'services'
+
+const qrAppearance = {
+  size: 240,
+  errorCorrectionLevel: 'high',
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js]: we have [config option for qr code](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/config.js#L38)
please use (update if necessary) it

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+              </Link>
+            </div>
+          </div>
+        </div>
+        <div className='content'>
+          <div className='container'>
+            <picture className='qr' />
+            <AddressPicker
+              addresses={items}
+              selectedAddress={selectedAddress}
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js]: I slightly updated AddressPicker API, for usage please refer to [this example](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/components/DigitalAssetSendForm/DigitalAssetSendForm.js#L54-L59)

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveViewContainer.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import {
+  openView,
+  closeView,
+} from './modules/digitalAssetsReceive'
+
+const mapStateToProps = (state: AppState) => {
+  const items = selectWalletsAddresses(state).addresses.map(address => ({ address, title: '' }))
+  return {
+    items,
+  }
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveViewContainer.js]: items should be AddressNames, [use correct selector](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/store/selectors/favorites.js#L59-L69)

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      <div className='digital-assets-receive-view'>
+        <div className='header'>
+          <div className='container'>
+            <div className='title'>
+              <JText value='Receive assets' size='header' color='gray' />
+            </div>
+            <div className='actions'>
+              <Link to='/digital-assets/'>
+                <JIcon name='cross' color='gray' />
+              </Link>
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js]: better to use `CloseableScreen` here with ability to use router back action

File: src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }
+}
+
+export function clean() {
+  return {
+    type: CLEAN,
+  }
+}
+
+export type DigitalAssetsManageAction = ExtractReturn<typeof openView> |
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: `DigitalAssetsManageAction ` -> `DigitalAssetsReceiveAction`

File: src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  ExtractReturn<typeof closeView> |
+  ExtractReturn<typeof setSearchQuery> |
+  ExtractReturn<typeof clean>
+
+const initialState: DigitalAssetsManageState = {
+  searchQuery: '',
+}
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: `DigitalAssetsManageState` -> `DigitalAssetsReceiveState`

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            </div>
+          </div>
+        </div>
+        <div className='content'>
+          <div className='container'>
+            <picture className='qr' />
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js]: We have [`QRCode`](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/routes/Funds/routes/Receive/components/FundsReceiveView.js#L26-L30) component, it supports overlay actions - copy/download.
I assume we should update it and integrate to this page

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js (hunk 1 above)
Comment (Creepypoke) [reply src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveView.js]: Yeah, I saw the config, but I thought, I was needed extend it, not update )

File: src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js [hunk 8]
```diff
@@ -19,50 +15,16 @@ export function closeView() {
...
-  ExtractReturn<typeof clean>
+export type DigitalAssetsReceiveAction = ExtractReturn<typeof openView> |
+  ExtractReturn<typeof closeView>
 
-const initialState: DigitalAssetsManageState = {
-  searchQuery: '',
-}
+const initialState: {} = {}
 
 const digitalAssetsReceive = (
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: but this reducer do nothing and could be easily removed

Comment (IViol) [reply src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: so this module will include only constants and actions - it's ok

Comment (IViol) [reply src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: also, as I see these actions are not used at all.
maybe this module could be removed entirely?

Comment (Creepypoke) [reply src/routes/DigitalAssets/routes/Receive/modules/digitalAssetsReceive.js]: Yeah, this unnecessary module was removed

File: src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveViewContainer.js [hunk 9]
```diff
@@ -3,18 +3,20 @@
...
 const mapStateToProps = (state: AppState) => {
-  const items: AddressNames = selectWalletsAddressesList(state)
-    .reduce((acc, address) => ({ ...acc, [address]: address }), {})
+  const walletsList: AddressNames = selectWalletsAddressesList(state)
+    .reduce((acc, address, index) => ({ ...acc, [address]: `Address #${index}` }), {})
+
+  const activeWallet = selectActiveWallet(state)
+  const walletAddress = activeWallet ? activeWallet.address : null
+  const items: AddressNames = isVoid(walletAddress) ?
+    walletsList : { [walletAddress]: 'Address #0' }
```
Comment (IViol) [inline src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveViewContainer.js]: on receive page we need to know several things:
1) current wallet
2) walletAddresses
3) addressNames map (to get names for addresses that was derived from mnemonic wallet)

`mapStateToProps` here will be very similar [to this one](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/layouts/MenuLayout/MenuLayoutContainer.js#L26-L43) [or another one](https://github.com/jibrelnetwork/jwallet-web/blob/develop/src/routes/Wallets/routes/Addresses/WalletsAddressesViewContainer.js#L24-L49)

I assume it's part of component's UI to decide which case to render (depends on `wallet.type`) - just one address for `address` wallet or selector with addresses (with names if they exist) for `mnemonic` wallet.

please check examples of usage of derived addresses from mnemonic

Comment (stcherenkov) [reply src/routes/DigitalAssets/routes/Receive/DigitalAssetsReceiveViewContainer.js]: Looks quite similar to me. Can you be more specific? What is wrong?

File: src/store/stateSelectors.js [hunk 10]
```diff
@@ -35,6 +35,10 @@ export function selectWalletsAddresses(state: AppState): WalletsAddressesState {
   return state.walletsAddresses
 }
 
+export function selectWalletsAddressesList(state: AppState): string[] {
+  return state.walletsAddresses.addresses
```
Comment (stcherenkov) [inline src/store/stateSelectors.js]: Property `state.walletsAddresses` always exists?

Comment (Creepypoke) [reply src/store/stateSelectors.js]: I don't know, really. We have a very level of entropy in selectors.

**Comment by Creepypoke** [comment]
@IViol Do you have any questions or comments more?

**Pull request merged**